internal/api/v1/upload.go

```diff
@@ -0,0 +1,94 @@
+package v1
+
+import (
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/epinio/epinio/helpers/tracelog"
+	"github.com/epinio/epinio/internal/cli/clients/gitea"
+	"github.com/julienschmidt/httprouter"
+	"github.com/mholt/archiver/v3"
+)
+
+// Upload receives the application data, as tarball, and creates the gitea as
+// well as k8s resources to trigger staging
+func (hc ApplicationsController) Upload(w http.ResponseWriter, r *http.Request) {
+	log := tracelog.Logger(r.Context())
+
+	params := httprouter.ParamsFromContext(r.Context())
+	org := params.ByName("org")
+	app := params.ByName("app")
+
+	log.Info("processing upload for", "org", org, "app", app)
+
+	gitea, err := gitea.New()
+	if err != nil {
+		internalError(w, err.Error())
+		return
+	}
+
+	log.V(2).Info("parsing multipart form")
+
+	file, _, err := r.FormFile("file")
+	if err != nil {
+		userError(w, "can't read multipart file input")
+		return
+	}
+	defer file.Close()
+
+	tmpDir, err := ioutil.TempDir("", "epinio-app")
+	if err != nil {
+		internalError(w, "can't create temp directory")
+		return
+	}
+	defer os.RemoveAll(tmpDir)
+
+	blob := path.Join(tmpDir, "blob.tar")
+	f, err := os.OpenFile(blob, os.O_WRONLY|os.O_CREATE, 0666)
+	if err != nil {
+		internalError(w, "failed create file for writing app sources to temp location")
+		return
+	}
+	defer f.Close()
+
+	_, err = io.Copy(f, file)
+	if err != nil {
+		internalError(w, "failed to copy app sources to temp location")
+		return
+	}
+
+	log.V(2).Info("unpacking temp dir")
+	appDir := path.Join(tmpDir, "app")
+	err = archiver.Unarchive(blob, appDir)
```

Comment:
To prevent bad actors from performing a DoS attacks on Epinio we should do such things in a queue in the background. Obviously that's not possible right now, because our storage solution (gitea) is supposed to store the code after extraction but that's something to keep in mind when we investigate #347 . I'll add this comment on that story too.

Reply:
This endpoint should not be accessible without authentication. I'm still hopeful, we can use a traefik middleware for that.

There is an option to validate the overall request size before reading to protect against uploads that are "too large". But I wasn't able to get it to work and I don't know what values (523mb?) to put. I think we should postpone that, I can create an issue so we don't forget though.

Regarding the unarchive, I'm not sure how expensive that is, the data is not compressed, so it's only IO and data in a tarball is ordered sequentially. Since this is a http upload, we need to move the data somewhere anyhow (read+insert/upload). The worker could then read the data again to write it to the final destination.

We could add a story to support "large" apps. I'm not sure what's considered large and how much resources that would effectively use.